### PR TITLE
[TKNECO-17] Golang S2I Task

### PR DESCRIPTION
**Pull request by Senjuti256**
Everything is running successfully.
   
In the pipelinerun I am using two tasks :- the git clone and this s2i task. I am passing a sample golang repository in the URL parameter of the git clone task. After finishing it the s2i go task starts getting executed and it will firstly generate the Dockerfile for the application source code in the generate step. The Dockerfile is stored in a workspace to be used by all the steps of the task. Using this Dockerfile the buildah tool will build the image and push it to a local container registry. After that the image digest will be displayed

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -7,14 +7,26 @@ BATS_CORE = ./test/.bats/bats-core/bin/bats
 BATS_FLAGS ?= --print-output-on-failure --show-output-of-passing-tests --verbose-run
 
 # path to the bats test files, overwite the variables below to tweak the test scope
-E2E_TESTS ?= ./test/e2e/*.bats
+E2E_TESTS ?= ./test/e2e/s2i-e2e.bats
```
Comment (otaviof) [inline Makefile:10]: Please revert the temporary change.

File: Makefile [hunk 2]
```diff
@@ -7,14 +7,26 @@ BATS_CORE = ./test/.bats/bats-core/bin/bats
...
+TASK_GIT ?= https://github.com/openshift-pipelines/task-git/releases/download/0.0.1/task-git-0.0.1.yaml
+
 
 # skopeo-copy task e2e test variables, source, destination url and tls-verify parameter.
 E2E_SC_PARAMS_SOURCE ?= docker://docker.io/library/busybox:latest
 E2E_SC_PARAMS_DESTINATION ?= docker://registry.registry.svc.cluster.local:32222/busybox:latest
 # setting tls-verify as false disables the HTTPS client as well, something we need for e2e testing
 E2E_SC_PARAMS_TLS_VERIFY ?= false
 
+E2E_SC_PARAMS_PATH_CONTEXT ?= .
```
Comment (otaviof) [inline Makefile:22]: Please add a top comment on every `Makefile` variable (or target).

File: Makefile [hunk 3]
```diff
@@ -25,10 +37,15 @@ E2E_BUILDAH_PVC_NAME ?= task-buildah
 # buildah task e2e test variables, image path, containerfile_path
 E2E_BUILDAH_CONTAINERFILE_PATH ?= /workspace/source/Dockerfile
 E2E_BUILDAH_IMAGE ?= test-buildah
-E2E_BUILDAH_REGISTRY ?= docker://registry.registry.svc.cluster.local:32222/test-buildah:latest
+E2E_BUILDAH_REGISTRY ?= registry.registry.svc.cluster.local:32222/test-buildah:latest
```
Comment (otaviof) [inline Makefile:40]: Let's rename this to reflect is target container registry, used for `buildah` tests. Also, we can reuse the registry part and compose afterwards.

Comment (otaviof) [reply Makefile:40]: On this case is also good to have a top comment with a complete description.

File: scripts/s2i-common.sh [hunk 4]
```diff
@@ -0,0 +1,46 @@
+#!/usr/bin/env bash
+
+declare -rx PARAMS_VERSION="${PARAMS_VERSION:-}"
+declare -rx PARAMS_PATH_CONTEXT="${PARAMS_PATH_CONTEXT:-}"
+declare -rx PARAMS_TLS_VERIFY="${PARAMS_TLS_VERIFY:-}"
+declare -rx PARAMS_VERBOSE="${PARAMS_VERBOSE:-}"
+
+#declare -rx RESULTS_IMAGE_DIGEST_PATH="${RESULTS_IMAGE_DIGEST_PATH:-}"
```
Comment (otaviof) [inline scripts/s2i-common.sh:8]: Let's not leave comments behind.

File: scripts/s2i-common.sh [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+#
+# Asserting Environment
+#
+
+declare -ra required_vars=(
+    PARAMS_VERSION
+    PARAMS_PATH_CONTEXT
+    #RESULTS_IMAGE_DIGEST_PATH
```
Comment (otaviof) [inline scripts/s2i-common.sh:17]: Also here.

File: scripts/s2i-prepare.sh [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+source "$(dirname ${BASH_SOURCE[0]})/s2i-common.sh"
+
+phase "Preparing the Workspaces, setting the expected ownership and permissions"
+#chmod -v 775 "${WORKSPACES_SOURCE_PATH}"
+
+if [[ "${WORKSPACES_SOURCE_BOUND}" == "true" ]]; then
+    echo "WORKSPACE EXISTS"
+else
+    fail "WORKSPACE_SOURCE is not bounded."
+fi
```
Comment (otaviof) [inline scripts/s2i-prepare.sh:16]: When the workspace is not bounded, the script needs to bail out.

File: scripts/s2i-results.sh [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+
+if [ -e "$filename" ]; then
+  echo "File exists: $filename"
+else
+  echo "File does not exist: $filename"
+  exit 1
+fi
+
+# Writing image digest to results
+cat "$filename" | tee /tekton/results/IMAGE_DIGEST
```
Comment (otaviof) [inline scripts/s2i-results.sh:20]: Let's initially leave this logic inside a single script. And assess later on.

File: templates/task-s2i-go.yaml [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+        - name: varlibcontainers
+          mountPath: /var/lib/containers
+        - name: gen-source
+          mountPath: /gen-source
+    - name: display-image-digest
+      image: busybox
+      command: ['cat', '/gen-source/image-digest']
+      volumeMounts:
+         - name: gen-source
+           mountPath: /gen-source
```
Comment (otaviof) [inline templates/task-s2i-go.yaml:89]: Let's split the scripts in a new pull-request...

File: templates/task-sc.yaml [hunk 9]
```diff
@@ -7,7 +7,7 @@ metadata:
     app.kubernetes.io/version: {{ .Chart.Version }}
 {{- if .Values.annotations }}
   annotations:
-  {{- .Values.annotations | toYaml | nindent 4 }}
+{{- .Values.annotations | toYaml | nindent 4 }}
```
Comment (otaviof) [inline templates/task-sc.yaml:10]: I recommend keeping the indentation block in place. TLDR; there are two entities going on, the actual YAML file, and the template logic thus deserves its own indentation level.

File: test.gen [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+COPY upload/src /tmp/src
+# Change file ownership to the assemble user. Builder image must support chown command.
+RUN chown -R 1001:0 /tmp/src
+USER 1001
+# Assemble script sourced from builder image based on user input or image metadata.
+# If this file does not exist in the image, the build will fail.
+RUN /usr/libexec/s2i/assemble
+# Run script sourced from builder image based on user input or image metadata.
+# If this file does not exist in the image, the build will fail.
+CMD /usr/libexec/s2i/run
```
Comment (otaviof) [inline test.gen:21]: It's okay to remove this :-)

File: test/e2e/resources/10-pipeline.yaml [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+      workspaces:
+        - name: output
+          workspace: source
+      params:
+        - name: URL
+          value: https://github.com/Senjuti256/goHelloWorld
+
+    - name: s2i-go-tr
+      taskRef:
+        name: s2i-go-tr
```
Comment (otaviof) [inline test/e2e/resources/10-pipeline.yaml:41]: nit: I'd name it as `s2i-golang-*`.

File: test/e2e/s2i-e2e.bats [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+
+
+
+# Spinning up a PipelineRun using the S2I Go Task to build and push a container image
+@test "[e2e] pipeline-run using s2i-go task" {
+  # Asserting required configuration is informed
+  [ -n "${E2E_PARAM_PATH_CONTEXT}" ]
+  [ -n "${E2E_PVC_NAME}" ]
+  [ -n "${E2E_S2I_IMAGE}" ]
+  [ -n "${E2E_S2I_TLS_VERIFY}" ] # Setting TLS_VERIFY=false
```
Comment (otaviof) [inline test/e2e/s2i-e2e.bats:18]:
```bash
[ "${E2E_S2I_TLS_VERIFY}" == "false" ]
```

File: test/e2e/s2i-e2e.bats [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+  # Cleaning up existing resources before starting a new pipelinerun
+  run kubectl delete pipelinerun --all
+  assert_success
+
+  # E2E PipelineRun
+  run tkn pipeline start s2i-go-tr \
+    --param="PATH_CONTEXT=${E2E_PARAM_PATH_CONTEXT}" \
+    --param="TLSVERIFY=${E2E_S2I_TLS_VERIFY}" \
+    --param="IMAGE=${E2E_S2I_IMAGE}" \
+		--param="VERBOSE=true" \
```
Comment (otaviof) [inline test/e2e/s2i-e2e.bats:30]: nit: please fix the indentation level.

File: test1.gen [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+COPY upload/src /tmp/src
+# Change file ownership to the assemble user. Builder image must support chown command.
+RUN chown -R 1001:0 /tmp/src
+USER 1001
+# Assemble script sourced from builder image based on user input or image metadata.
+# If this file does not exist in the image, the build will fail.
+RUN /usr/libexec/s2i/assemble
+# Run script sourced from builder image based on user input or image metadata.
+# If this file does not exist in the image, the build will fail.
+CMD /usr/libexec/s2i/run
```
Comment (otaviof) [inline test1.gen:21]: Please remove the temporary file.

File: test2.gen [hunk 15]
```diff
@@ -0,0 +1,21 @@
...
+COPY upload/src /tmp/src
+# Change file ownership to the assemble user. Builder image must support chown command.
+RUN chown -R 1001:0 /tmp/src
+USER 1001
+# Assemble script sourced from builder image based on user input or image metadata.
+# If this file does not exist in the image, the build will fail.
+RUN /usr/libexec/s2i/assemble
+# Run script sourced from builder image based on user input or image metadata.
+# If this file does not exist in the image, the build will fail.
+CMD /usr/libexec/s2i/run
```
Comment (otaviof) [inline test2.gen:21]: And this one too.

**Review by otaviof — CHANGES_REQUESTED** [review]
Good overall progress, let's work on the review feedback.

File: Makefile (hunk 3 above)
Comment (Senjuti256) [reply Makefile:40]: This one was done by Anand so I did not touch it fearing any loss of functionality or malfunctionality in his task on editing it.

File: test/e2e/resources/10-pipeline.yaml [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+  tasks:
+    - name: git
+      taskRef:
+        name: git
+      workspaces:
+        - name: output
+          workspace: source
+      params:
+        - name: URL
+          value: https://github.com/Senjuti256/goHelloWorld
```
Comment (otaviof) [inline test/e2e/resources/10-pipeline.yaml:37]: Could we use a common sample, as in any s2i official Golang sample repository would be great for this use-case.

File: test/e2e/s2i-e2e.bats [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+
+
+
+# Spinning up a PipelineRun using the S2I Go Task to build and push a container image
+@test "[e2e] pipeline-run using s2i-go task" {
+  # Asserting required configuration is informed
+  [ -n "${E2E_PARAM_PATH_CONTEXT}" ]
+  [ -n "${E2E_PVC_NAME}" ]
+  [ -n "${E2E_S2I_IMAGE}" ]
+  [ -n "${E2E_S2I_TLS_VERIFY}" ]
```
Comment (otaviof) [inline test/e2e/s2i-e2e.bats:18]: Is it required to be set to `false` during testing?

File: values.yaml [hunk 18]
```diff
@@ -3,6 +3,8 @@ images:
   bash: docker.io/library/bash:latest
   skopeo: quay.io/skopeo/stable:latest
   buildah: registry.access.redhat.com/ubi8/buildah:latest
+  s2igo: quay.io/openshift-pipeline/s2i:latest
+  s2ibuildah: quay.io/buildah/stable:latest
```
Comment (otaviof) [inline values.yaml:7]: It should be using the same `buildah` image referenced above.

**Review by otaviof — CHANGES_REQUESTED** [review]
Cool, very good progress! Some more remarks 🙏

File: test/e2e/s2i-e2e.bats (hunk 17 above)
Comment (Senjuti256) [reply test/e2e/s2i-e2e.bats:18]: Local container registry cannot be used during the e2e testing unless the tls-verification is set to false. So i think it's necessary.

File: values.yaml (hunk 18 above)
Comment (Senjuti256) [reply values.yaml:7]: Actually here the s2igo and s2ibuildah images are redundant. I mentioned them here keeping in mind that in future if I separate the commands used in the s2i go task as separate scripts. Should I remove them?

File: test/e2e/resources/10-pipeline.yaml (hunk 16 above)
Comment (Senjuti256) [reply test/e2e/resources/10-pipeline.yaml:37]: I am looking for some sample repository and updating it.

Comment (Senjuti256) [reply test/e2e/resources/10-pipeline.yaml]: I have updated with a sample repository.

File: .github/workflows/test.yaml [hunk 19]
```diff
@@ -34,7 +34,8 @@ jobs:
         with:
           pipeline_version: v0.41.0
           cli_version: v0.29.1
-
```
Comment (otaviof) [inline .github/workflows/test.yaml:37]: nit: please revert the changes on this file.

File: Makefile [hunk 20]
```diff
@@ -25,10 +36,15 @@ E2E_BUILDAH_PVC_NAME ?= task-buildah
...
 E2E_BUILDAH_TLS_VERIFY ?= false
 E2E_BUILDAH_POPULATE_WORKSPACE ?= test/e2e/resources/populate-workspace-task.yaml
 
+#The local container registry to push the image during e2e testing of s2i-golang task
+E2E_S2I_IMAGE ?= registry.registry.svc.cluster.local:32222/test-s2i:latest
+E2E_S2I_TLS_VERIFY ?= false
```
Comment (otaviof) [inline Makefile:45]: Please, add a top comment explaining why TLS verification needs to be false in combination with the internal container registry (during end-to-end).

File: templates/task-s2i-go.yaml [hunk 21]
```diff
@@ -0,0 +1,94 @@
...
+    s2i-go task clones a Git repository and builds and
+    pushes a container image using S2I and a Go builder image.
+
+  results:
+    - name: IMAGE_DIGEST
+      description: Digest of the image just built.
+  params:
+    - name: VERSION
+      description: The tag of go imagestream for go version
+      default: '1.14.7-ubi8'
```
Comment (otaviof) [inline templates/task-s2i-go.yaml:22]: It should become a `values.yaml` entry.

File: templates/task-s2i-go.yaml [hunk 22]
```diff
@@ -0,0 +1,94 @@
...
+      type: string
+    - name: PATH_CONTEXT
+      description: The location of the path to run s2i from.
+      default: .
+      type: string
+    - name: TLSVERIFY
```
Comment (otaviof) [inline templates/task-s2i-go.yaml:28]: Please rename this variable to `TLS_VERIFY` following the other Tasks.

File: templates/task-s2i-go.yaml [hunk 23]
```diff
@@ -0,0 +1,94 @@
...
+      volumeMounts:
+        - name: gen-source
+          mountPath: /gen-source
+    - name: build
+      image: quay.io/buildah/stable:latest
+      workingDir: /gen-source
+      command: ['buildah', 'bud', '--storage-driver=vfs', '--tls-verify=$(params.TLSVERIFY)', '--layers', '-f', '/gen-source/Dockerfile.gen', '-t', '$(params.IMAGE)', '.']
+      env:
+        - name: GO111MODULE
+          value: "auto"
```
Comment (otaviof) [inline templates/task-s2i-go.yaml:69]: That's not applicable to `buildah` step.

File: templates/task-s2i-go.yaml [hunk 24]
```diff
@@ -0,0 +1,94 @@
...
+        - name: varlibcontainers
+          mountPath: /var/lib/containers
+        - name: gen-source
+          mountPath: /gen-source
+    - name: display-image-digest
+      image: busybox
+      command: ['cat', '/gen-source/image-digest']
+      volumeMounts:
+         - name: gen-source
+           mountPath: /gen-source
```
Comment (otaviof) [inline templates/task-s2i-go.yaml:89]: What we want to do is use the `buildah` Task to compose a Pipeline. So, we need to remove those steps and try to use that Task instead.

File: test/.bats/bats-core
Comment (otaviof) [inline test/.bats/bats-core:1]: Please, lets not update submodules on this PR.

File: test/.bats/bats-file
Comment (otaviof) [inline test/.bats/bats-file:1]: Please, let's not update submodules on this PR.

File: test/e2e/resources/10-pipeline.yaml [hunk 25]
```diff
@@ -0,0 +1,58 @@
...
+      taskRef:
+        name: git
+      workspaces:
+        - name: output
+          workspace: source
+      params:
+        - name: URL
+          value: https://github.com/hackersandslackers/golang-helloworld      #This will basically contain the reference to the github repository containing the Go application source code whose image we want to generate.
+        - name: REVISION
+          value: master
```
Comment (otaviof) [inline test/e2e/resources/10-pipeline.yaml:39]: Lovely 👍

File: test/e2e/s2i-e2e.bats (hunk 17 above)
Comment (otaviof) [reply test/e2e/s2i-e2e.bats:18]: Cool, this goes along with the Makefile's "top comment".

File: upload/src
Comment (otaviof) [inline upload/src:1]: Please let's not update submodules on this PR.

File: values.yaml [hunk 26]
```diff
@@ -3,7 +3,6 @@ images:
   bash: docker.io/library/bash:latest
   skopeo: quay.io/skopeo/stable:latest
   buildah: registry.access.redhat.com/ubi8/buildah:latest
-
```
Comment (otaviof) [inline values.yaml:6]: nit: please leave the empty line, they should serve as a logical block separation.

**Review by otaviof — CHANGES_REQUESTED** [review]
Great progress, some last comments.

File: .github/workflows/test.yaml (hunk 19 above)
Comment (Senjuti256) [reply .github/workflows/test.yaml:37]: Done.

File: test/.bats/bats-core
Comment (Senjuti256) [reply test/.bats/bats-core:1]: How to drop the updates in the submodules?

**Comment by otaviof** [comment]
Closing this PR in favor of https://github.com/openshift-pipelines/task-containers/pull/17.

**Comment by otaviof** [comment]
Closing this PR in favor of https://github.com/openshift-pipelines/task-containers/pull/17.

**Comment by otaviof** [comment]
Closing this PR in favor of https://github.com/openshift-pipelines/task-containers/pull/17.
